stdlib/json.gr

```diff
@@ -0,0 +1,2103 @@
+include "runtime/unsafe/tags"
+include "runtime/unsafe/wasmi32"
+include "runtime/unsafe/wasmi64"
+include "runtime/unsafe/wasmf64"
+include "runtime/atof/parse" as Atof
+include "buffer"
+include "char"
+include "string"
+include "list"
+include "uint8"
+from RuntimeString use { toString as runtimeToString, getCodePoint }
+from Numbers use { coerceNumberToWasmI32 }
+from DataStructures use { tagSimpleNumber, untagSimpleNumber }
+
+// Primitive offsets
+// TODO(#703): Get these offsets from the runtime
+@unsafe
+let _INT64_BOXED_VALUE_OFFSET = 8n
+@unsafe
+let _Float64_BOXED_VALUE_OFFSET = 8n
+
+/**
+ * Data structure representing the result of parsing and the input of printing JSON.
+ *
+ * This data structure is semantically equivalent to the JSON format allowing
+ * mostly lossless round trips of printing and parsing. Exceptions to this are
+ * whitespace, multiple ways JSON allows escaping characters in strings, and
+ * some edge cases related to Grain's `Number` type.
+ *
+ * @example
+ * Json.parse("{\"currency\":\"€\",\"price\":99.99}") == JsonObject([
+ *   ("currency", JsonString("€")),
+ *   ("price", JsonNumber(99.99)),
+ * ])
+ *
+ * @example
+ * Json.parse("{\n\"currency\":\"€\",\n\"price\":99.99\n}") == JsonObject([
+ *   ("currency", JsonString("€")),
+ *   ("price", JsonNumber(99.99)),
+ * ])
+ */
+provide enum rec Json {
+  JsonNull,
+  JsonBoolean(Bool),
+  JsonNumber(Number),
+  JsonString(String),
+  JsonArray(List<Json>),
+  // Note that JsonObject here is deliberately defined as a simple list of key value pair tuples as opposed
+  // to for example a Map in order to accommodate the fact that the ECMA-404 standard doesn't prohibit
+  // duplicate names in Objects. Such JSON should be representable by the JSON data structure for lossless
+  // processing. This also simplifies implementation by not requiring a purpose built data structure and
+  // has the benefit of List's immutability. It's a conscious decision that sacrifices ease of use of the
+  // API for lossless handing of these edge cases with intention of later building more ergonomic APIs on a
+  // higher level of abstraction.
+  JsonObject(List<(String, Json)>),
+}
+
+// TODO(#992): Remove part about occurrence here because it is covered by the variant doc.
+/**
+ * Represents errors for cases where a `JSON` object cannot be represented in the
+ * JSON format along with a human readable text message. This can happen when
+ * it contains number values `NaN`, `Infinity` or `-Infinity`.
+ */
+provide enum JsonToStringError {
+  /**
+   * The JSON object contains a number value of `NaN`, `Infinity` or `-Infinity`.
+   */
+  InvalidNumber(String),
+}
+
+/**
+ * Controls how indentation is performed in custom formatting.
+ *
+ * Following examples have whitespaces and line breaks replaced with visible
+ * characters for illustrative purposes.
+ *
+ * `NoIndentation`
+ * ```
+ * {↵
+ * "currency":·"€",↵
+ * "price":·99.9↵
+ * }
+ * ```
```

Comment:
These will be moved to graindoc descriptions when it is supported by the PR linked in a previous comment.